rustc_hir_analysis: add a helper to check function the signature mismatches

eduardosm opened the pull request:
This function is now used to check `#[panic_handler]`, `start` lang item, `main`, `#[start]` and intrinsic functions.

The diagnosis produced are now closer to the ones produced by trait/impl method signature mismatch.

This is the first time I do anything with rustc_hir_analysis/rustc_hir_typeck, so comments and suggestions about things I did wrong or that could be improved will be appreciated.

compiler-errors commented on tests/ui/panic-handler/panic-handler-bad-signature-2.stderr:
```diff
@@ -1,8 +1,18 @@
+   |
+   = note: expected signature `fn(&PanicInfo<'_>) -> _`
+              found signature `fn(&'static PanicInfo<'_>) -> _`
+note: the anonymous lifetime as defined here...
```
This is pretty confusing. We probably want to actually replace the call to `ocx.resolve_regions_and_report_errors` to a manual call to `ocx.infcx.resolve_regions` and report any region errors manually from there with a more descriptive message.

eduardosm replied:
What message would we want here?

In the case of a trait/impl:
```rust
use std::panic::PanicInfo;

trait Tr {
    fn panic(info: &PanicInfo<'_>);
}

impl Tr for () {
    fn panic(_info: &'static PanicInfo) {
        unimplemented!();
    }
}
```
It is reported similarly:
```
error[E0308]: method not compatible with trait
 --> lib.rs:8:5
  |
8 |     fn panic(_info: &'static PanicInfo) {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ lifetime mismatch
  |
  = note: expected signature `fn(&PanicInfo<'_>)`
             found signature `fn(&'static PanicInfo<'_>)`
note: the anonymous lifetime as defined here...
 --> lib.rs:8:30
  |
8 |     fn panic(_info: &'static PanicInfo) {
  |                              ^^^^^^^^^
  = note: ...does not necessarily outlive the static lifetime

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

The difference is that in the test, the note spans the whole signature instead of just `PanicInfo`:
```
note: the anonymous lifetime as defined here...
  --> $DIR/panic-handler-bad-signature-2.rs:9:1
   |
LL | fn panic(info: &'static PanicInfo) -> !
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: ...does not necessarily outlive the static lifetime
```
I'm not sure why.

eduardosm replied:
I opened https://github.com/rust-lang/rust/issues/115903 regarding confusing lifetime diagnosis for trait impl functions. I think fixing that its outside the scope of this PR.

compiler-errors commented on compiler/rustc_hir_analysis/src/check/mod.rs:
```diff
@@ -546,3 +554,80 @@ fn bad_non_zero_sized_fields<'tcx>(
+pub fn check_function_signature<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    cause_code: ObligationCauseCode<'tcx>,
+    fn_id: LocalDefId,
```
This can be a `DefId`, and you can use `CRATE_DEF_ID` for the cases that it's required to be a `LocalDefId`.

compiler-errors commented on compiler/rustc_hir_analysis/src/check/mod.rs:
```diff
@@ -546,3 +554,80 @@ fn bad_non_zero_sized_fields<'tcx>(
+
+pub fn check_function_signature<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    cause_code: ObligationCauseCode<'tcx>,
```
Can you make the caller pass in a full `ObligationCause`? That's usually more conventional than just passing the code.

compiler-errors commented on compiler/rustc_hir_analysis/src/check/mod.rs:
```diff
@@ -546,3 +554,80 @@ fn bad_non_zero_sized_fields<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    cause_code: ObligationCauseCode<'tcx>,
+    fn_id: LocalDefId,
+    expected_sig: ty::PolyFnSig<'tcx>,
```
This should always just be a `FnSig`, is there ever a case where we're not passing `ty::Binder::dummy`?

eduardosm replied:
We use `ty::Binder::bind_with_vars` for intrinsics (I guess only needed for those with lifetime params) and for the panic handler.

compiler-errors replied:
Yeah, ok

compiler-errors commented on compiler/rustc_hir_analysis/src/check/mod.rs:
```diff
@@ -546,3 +554,80 @@ fn bad_non_zero_sized_fields<'tcx>(
+
+    let expected_sig = tcx.liberate_late_bound_regions(fn_id.into(), expected_sig);
+
+    match ocx.sup(&cause, param_env, expected_sig, actual_sig) {
```
It concerns me that we're only checking the subtyping in one direction instead of both directions (with `eq` here).

eduardosm commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #115909) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

eduardosm committed:
rustc_hir_analysis: add a helper to check function the signature mismatches

This function is now used to check `#[panic_handler]`, `start` lang item, `main`, `#[start]` and intrinsic functions.

The diagnosis produced are now closer to the ones produced by trait/impl method signature mismatch.

compiler-errors commented on compiler/rustc_hir_analysis/src/check/mod.rs:
```diff
@@ -546,3 +554,82 @@ fn bad_non_zero_sized_fields<'tcx>(
+
+    let expected_sig = tcx.liberate_late_bound_regions(fn_id, expected_sig);
+
+    match ocx.eq(&cause, param_env, expected_sig, actual_sig) {
```
This is not how to compare two binders. This is not very much different than using `sub`, which I think is still wrong. You should probably wrap both binders in a `FnPtr` and compare them directly. The error reporting will kinda suck, but that can be dealt with differently.

eduardosm replied:
The output changed from 
```
   = note: expected signature `fn(&PanicInfo<'_>) -> _`
              found signature `fn(&'static PanicInfo<'_>) -> _`
```
to
```
   = note: expected fn pointer `for<'a, 'b> fn(&'a PanicInfo<'b>) -> _`
              found fn pointer `for<'a> fn(&'static PanicInfo<'a>) -> _`
```

But only for lifetime errors.

For non-lifetime errors, it stayed expected/found signature, I believe that thanks to passing the signatures in `infer::ValuePairs` here (I'm not sure that's entirely correct)
```rust
err_ctxt.note_type_err(
    &mut diag,
    &cause,
    None,
    Some(infer::ValuePairs::Sigs(ExpectedFound {
        expected: tcx.liberate_late_bound_regions(fn_id, expected_sig),
        found: tcx.liberate_late_bound_regions(fn_id, actual_sig),
    })),
    err,
    false,
    false,
);
```

bors commented:
:umbrella: The latest upstream changes (presumably #115952) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

eduardosm committed:
wrap fn sig binders in fn ptr

bors commented:
:pushpin: Commit 85d61b01aea90d9201c35c4bd61a508c72807c92 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 85d61b01aea90d9201c35c4bd61a508c72807c92 -->
<!-- homu: {"type":"Approved","sha":"85d61b01aea90d9201c35c4bd61a508c72807c92","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 85d61b01aea90d9201c35c4bd61a508c72807c92 with merge 99b63d068b0be31b4d8a28f93921b5d14f13cdc0...
<!-- homu: {"type":"BuildStarted","head_sha":"85d61b01aea90d9201c35c4bd61a508c72807c92","merge_sha":"99b63d068b0be31b4d8a28f93921b5d14f13cdc0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6267951093/job/17023972889)
Approved by: compiler-errors
Pushing 99b63d068b0be31b4d8a28f93921b5d14f13cdc0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6267951093/job/17023972889"},"merge_sha":"99b63d068b0be31b4d8a28f93921b5d14f13cdc0"} -->

rust-timer commented:
Finished benchmarking commit (99b63d068b0be31b4d8a28f93921b5d14f13cdc0): [comparison URL](https://perf.rust-lang.org/compare.html?start=b3aa8e7168a3d940122db3561289ffbf3f587262&end=99b63d068b0be31b4d8a28f93921b5d14f13cdc0&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b3aa8e7168a3d940122db3561289ffbf3f587262&end=99b63d068b0be31b4d8a28f93921b5d14f13cdc0&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b3aa8e7168a3d940122db3561289ffbf3f587262&end=99b63d068b0be31b4d8a28f93921b5d14f13cdc0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.7%, 0.7%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.0% | [1.5%, 2.6%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.7%, 0.7%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b3aa8e7168a3d940122db3561289ffbf3f587262&end=99b63d068b0be31b4d8a28f93921b5d14f13cdc0&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=b3aa8e7168a3d940122db3561289ffbf3f587262&end=99b63d068b0be31b4d8a28f93921b5d14f13cdc0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 634.138s -> 633.396s (-0.12%)
**Artifact size**: 317.56 MiB -> 317.61 MiB (0.02%)
